amaranth/hdl/_ir.py

```diff
@@ -372,208 +349,137 @@ def _propagate_domains(self, missing_domain, *, platform=None):
+                raise NameError("Instance keyword argument {}={!r} does not start with one of "
+                                "\"a_\", \"p_\", \"i_\", \"o_\", or \"io_\""
+                                .format(kw, arg))
 
-    def prepare(self, ports=None, missing_domain=lambda name: _cd.ClockDomain(name)):
-        from ._xfrm import DomainLowerer
 
-        new_domains = self._propagate_domains(missing_domain)
-        fragment = DomainLowerer()(self)
-        if ports is None:
-            fragment._propagate_ports(ports=(), all_undef_as_ports=True)
-        else:
-            if not isinstance(ports, tuple) and not isinstance(ports, list):
-                msg = "`ports` must be either a list or a tuple, not {!r}"\
-                        .format(ports)
-                if isinstance(ports, _ast.Value):
-                    msg += " (did you mean `ports=(<signal>,)`, rather than `ports=<signal>`?)"
-                raise TypeError(msg)
-            mapped_ports = []
-            # Lower late bound signals like ClockSignal() to ports.
-            port_lowerer = DomainLowerer(fragment.domains)
-            for port in ports:
-                if not isinstance(port, (_ast.Signal, _ast.ClockSignal, _ast.ResetSignal)):
-                    raise TypeError("Only signals may be added as ports, not {!r}"
-                                    .format(port))
-                mapped_ports.append(port_lowerer.on_value(port))
-            # Add ports for all newly created missing clock domains, since not doing so defeats
-            # the purpose of domain auto-creation. (It's possible to refer to these ports before
-            # the domain actually exists through late binding, but it's inconvenient.)
-            for cd in new_domains:
-                mapped_ports.append(cd.clk)
-                if cd.rst is not None:
-                    mapped_ports.append(cd.rst)
-            fragment._propagate_ports(ports=mapped_ports, all_undef_as_ports=False)
-        return fragment
-
-    def _assign_names_to_signals(self):
-        """Assign names to signals used in this fragment.
+class Design:
+    """Represents a design ready for simulation or netlist building.
 
-        Returns
-        -------
-        SignalDict of Signal to str
-            A mapping from signals used in this fragment to their local names. Because names are
-            deduplicated using local information only, the same signal used in a different fragment
-            may get a different name.
+    Returned by ``Fragment.prepare``."""
+
+    def __init__(self, fragment, ports, *, hierarchy):
+        self.fragment = fragment
+        self.ports = ports
+        self.hierarchy = hierarchy
+        # dict of Fragment to SignalDict of Signal to name
+        self.signal_names = {}
+        self.fragment_names = {}
+        self._assign_names_to_signals(fragment, ports)
+        self._assign_names_to_fragments(fragment, hierarchy)
+        self._fixup_port_names()
```

Comment:
Something like this?

    self.ports = [(name or top_names[signal], signal, dir) for (name, signal, dir) in self.ports]